### DISCUSSION TITLE
Integration test with an app using Drift + provider

### DISCUSSION THREAD

**Discussion by frankvollebregt** [simolus3/drift#2534] [Q&A]
I'm currently (and long overdue) setting up integration tests for my app. I am providing, amongst some other utilities, my `drift` database via a `MultiProvider`. When I run a test that just loads the first screen (a login view), I can `find` Widgets just fine, but when I rebuild the screen (to see the results of interactions), using `widgetTester.pump` or `widgetTester.pumpAndSettle`, The following message is printed and the test fails:
```
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following ConnectionClosedException was thrown running a test:
Instance of 'ConnectionClosedException'

When the exception was thrown, this was the stack:
#0      DriftCommunication.request (package:drift/src/remote/communication.dart:113:66)
#1      _RemoteQueryExecutor.ensureOpen (package:drift/src/remote/client_impl.dart:158:10)
#2      DatabaseConnection.ensureOpen (package:drift/src/runtime/api/connection.dart:87:63)
#3      LazyDatabase.ensureOpen.<anonymous closure> (package:drift/src/utils/lazy_database.dart:61:49)
<asynchronous suspension>
<asynchronous suspension>
(elided 6 frames from dart:async and package:stack_trace)
```
I'm using my database from both the main isolate as well as a background isolate (for synchronization). In order to achieve this, I use the following code to launch and connect this isolate (which may have something to do with the problem I'm encountering...):
```dart
  static Future<QueryExecutor> _getBackgroundExecutor(File file) async {
    // Either get the existing connect port, or create a new connection
    final existingPort = IsolateNameServer.lookupPortByName(databaseIsolate);

    if (existingPort != null) {
      // We can connect to the existing database isolate!
      return DriftIsolate.fromConnectPort(existingPort).connect();
    } else {
      // We need to spawn a new isolate and register it
      final isolate = DriftIsolate.inCurrent(() => NativeDatabase(file));

      IsolateNameServer.registerPortWithName(
        isolate.connectPort,
        databaseIsolate,
      );

      return isolate.connect();
    }
  }

  static LazyDatabase _usableInBackground() {
    return LazyDatabase(() async {
      final dbFolder = await getApplicationDocumentsDirectory();
      final file = File(p.join(dbFolder.path, 'db.sqlite'));
      return _getBackgroundExecutor(file);
    });
  }
```
And this is what my tests look like:
```dart
void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  group('end-to-end test', () {
    testWidgets('Find the button (this works)', (tester) async {
      await Firebase.initializeApp();

      await tester.pumpWidget(const ProviderWrapper(child: MyApp()));

      final loginButton = find.byType(OutlinedButton);

      // Find the login button
      expect(loginButton, findsOneWidget);

      // There is no error message yet
      expect(find.text('This email address is invalid'), findsNothing);
    });

    testWidgets('Rebuild the Widget (this does not work)', (tester) async {
      await Firebase.initializeApp();

      await tester.pumpWidget(const ProviderWrapper(child: MyApp()));

      final loginButton = find.byType(OutlinedButton);

      // Find the login button
      expect(loginButton, findsOneWidget);

      // There is no error message yet
      expect(find.text('This email address is invalid'), findsNothing);

      // This causes the test to fail!
      await tester.pump();  // same with await tester.pumpAndSettle()
    });
  });
}
```
Does any of the above clearly signify a mistake of some sort? Is there any additional setup that I need to do in order to use my Drift database in an integration test?

Thanks!

**Comment by simolus3** [comment]
The drift isolate should outlive an individual test, right? Both snippets look reasonable to me. I've heard some other reports about the connection closed exceptions. The stack trace here is not very helpful unfortunately. But I don't think it's coming from a stray async task still using the old database. The request in `_RemoteQueryExecutor` is only sent once when opening the database.

Do you have a way to explicitly close the database class (not the drift isolate instance) in the test? If you only have one test, but explicitly close the database at the end and add a pump and settle with a short duration, do you get the same error?

I'll see if I can reproduce this with a similar setup tomorrow - I don't see anything obvious that could be wrong with these snippets.

**Reply by simolus3** [reply]
I couldn't reproduce this with [this simple app](https://github.com/simolus3/drift/tree/reproduce-2534/drift_isolate_repro).

I'm using a provider to setup and tear down the database, and it's using the isolate name server to share a drift isolate as well. Both tests try to open the database and run a simple query, which works in both of them. It's probably not easy to detect since your app is way more complex, but do you see any major difference in the way I'm setting things up in that repro?
